### ISSUE TITLE
problems building with LLVM 4.0 on macOS

### ISSUE THREAD

**Issue by pmetzger** [ruby-llvm/ruby-llvm#53]
On macOS, having installed LLVM 4.0 using macports, if I
run "rake", in ext/ruby-llvm-support/ I get

`ld: library not found for -lLLVM-4.0`

and the build dies.

I got past that in a cheesy way (I cut and pasted the g++ build
line and hand-edited -lLLVM-4.0 to -lLLVM) and then I used otool to
check that the resulting libRubyLLVMSupport-4.0.dylib was pointing to
the correct libLLVM (it is).

I then went off to run "rake generate_ffi" and got:

> rake aborted!
> LoadError: Could not open library 'libclang-3.5.so.1': dlopen(libclang-3.5.so.1, 5): image not found.
> Could not open library 'libclang-3.5.so.1.dylib': dlopen(libclang-3.5.so.1.dylib, 5): image not found.
> Could not open library 'libclang.so.1': dlopen(libclang.so.1, 5): image not found.
> Could not open library 'libclang.so.1.dylib': dlopen(libclang.so.1.dylib, 5): image not found.
> Could not open library 'clang': dlopen(clang, 5): image not found.
> Could not open library 'libclang.dylib': dlopen(libclang.dylib, 5): image not found

This seemed rather weird to me, since the string "3.5" doesn't appear
anywhere in the ruby-llvm sources I have.

**Comment by whitequark** [comment]
> run "rake", in ext/ruby-llvm-support/ I get
> 
> ld: library not found for -lLLVM-4.0
> 
> and the build dies.

It's mildly surprising that this broke. If you look at https://github.com/ruby-llvm/ruby-llvm/blob/master/ext/ruby-llvm-support/Rakefile#L29-L31, we already try to detect both the un-prefixed and the prefixed version. The only way it'll break that I see is if macports installs libLLVM.so but *not* llvm-config, which is decidedly a wrong thing to do. Or maybe the logic is borked and I don't see that.

> LoadError: Could not open library 'libclang-3.5.so.1': dlopen(libclang-3.5.so.1, 5): image not found.

The libclang-3.5 dependency comes from the ffi_gen gem. You'll find that its bindings are fairly similar to this library, and can be updated in much the same ways. I can also accept PRs there. (Or you can just use clang-3.5 for development for time being...)

**Comment by pmetzger** [comment]
So, macports installs llvm-config. I'll play with it and see if I can figure out what might be wrong with the logic.

BTW, is there a reason that version.rb doesn't try using llvm-config to get a version number?

**Comment by whitequark** [comment]
Because we only support a single version at a time anyway.

**Comment by thoughtafter** [comment]
I've updated master to LLVM 4.0 and created a llvm-4.0 branch to use. This should mean you no longer have to run rake generate_ffi yourself.

**Comment by pmetzger** [comment]
Thank you, but I couldn't afford to lose so much working time, so I abandoned my Ruby based prototype a while ago when it became apparent that I wasn't going to get any real help fixing it myself. Given that this took over a month, that was probably the right decision for me.

Apologies for not closing out my tickets, I should have. I really really really wanted to use ruby for my work, and perhaps I'll try ruby-llvm again in the future, but for now I've moved on.
